PyQt5-stubs/sip.pyi

```diff
@@ -40,12 +40,16 @@ class wrapper(simplewrapper): ...
 Buffer = Union['array', 'voidptr', str, bytes, bytearray]
 
 
+C = TypeVar("C")
 T = TypeVar("T")
 
 
 # The array type.
 class array(Sequence[T]):
-    def __getitem__(self, key: Any) -> Any: ...
+    @overload
+    def __getitem__(self, key: int) -> T: ...
+    @overload
+    def __getitem__(self: C, key: slice) -> C: ...
```

Comment:
Shouldn't the slice one be
```python
def __getitem__(self , key: slice) -> array[T]: ...
```

Reply:
I think the generic `self` is for returning a modified version of the same object

Reply:
Isn't this more flexible for inheritance while still hinting the same thing for a `sip.array` itself?  But sure, even if I am correct, that is at least unlikely to matter.

Reply:
Yeah, I'm not too sure how to make covariant return typing with something that's already Generic

Reply:
So typeshed just uses covariant all over `typing.Sequence`.

https://github.com/python/typeshed/blob/734d91f90bee3c782aa6c7c04492882dbc3c47b1/stdlib/3/typing.pyi#L292-L298
```python
class Sequence(_Collection[_T_co], Reversible[_T_co], Generic[_T_co]):
    @overload
    @abstractmethod
    def __getitem__(self, i: int) -> _T_co: ...
    @overload
    @abstractmethod
    def __getitem__(self, s: slice) -> Sequence[_T_co]: ...
```

I'm still taking passes trying to get variance internalized well.  I suppose I'll just go with that (minus abstract and with `array` not `Sequence`, but I am curious how `C` would get it wrong.  I would expect it to pass everything through.  Any chance you have any reference for the idea that the `C` approach relates to returning `self`?

Reply:
This is what I'm referencing when I say "generic self":
https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self

Reply:
Actually, looking at that page again, I think it actually aligns better with what you're saying about covariance. Maybe the `-> C` is better?

Reply:
Thanks.

> Note that when overriding a method with generic self, you must either return a generic self too, or return an instance of the current class. In the latter case, you must implement this method in all future subclasses.

I don't get why the `C` wouldn't work with the method being inherited.  But regardless, they said it's basically a questionable implementation presently so may as well not tempt fate.  I'm fine using the typeshed form.

Reply:
By that, you mean the current implementation?

Reply:
Yes.

Reply:
Oh I meant to ask this earlier. Do we need to annotate `__setitem__`?

Reply:
I guess we could fill out the `help()` entries anyways.

```
class array(builtins.object)
 |  Methods defined here:
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __getitem__(self, key, /)
 |      Return self[key].
 |  
 |  __len__(self, /)
 |      Return len(self).
 |  
 |  __setitem__(self, key, value, /)
 |      Set self[key] to value.
```

Reply:
Gah not trying to make this PR drag on forever... but there's also a `slice` overload for `__setitem__` to deal with 😖

Reply:
You know... I _did_ think about that...  just enough to forget it.  And yeah, I knew when I put five changes in one that something would hold them up.  `:]`